tests/client/core/test_middleware.js

```diff
@@ -0,0 +1,93 @@
+import { prefixMiddleWare } from 'core/middleware';
+
+describe('Prefix Middleware', () => {
+  let fakeRes;
+  let fakeNext;
+  let fakeConfig;
+
+  beforeEach(() => {
+    fakeNext = sinon.stub();
+    fakeRes = {
+      locals: {},
+      redirect: sinon.stub(),
+      set: sinon.stub(),
+    };
+    fakeConfig = new Map();
+    fakeConfig.set('validClientApplications', ['firefox', 'android']);
+  });
+
+  it('should call res.redirect', () => {
+    const fakeReq = {
+      originalUrl: '/en-us/firefox',
+      headers: {},
+    };
+    prefixMiddleWare(fakeReq, fakeRes, fakeNext, { _config: fakeConfig });
+    assert.ok(fakeRes.redirect.called);
+    assert.notOk(fakeNext.called);
+  });
+
+  it('should add an application when missing', () => {
+    const fakeReq = {
+      originalUrl: '/en-US/whatever/',
+      headers: {},
+    };
+    prefixMiddleWare(fakeReq, fakeRes, fakeNext, { _config: fakeConfig });
+    assert.equal(fakeRes.redirect.getCall(0).args[1], '/en-US/firefox/whatever/');
+  });
+
+  it('should add a lang when missing and leave a valid app first part intact', () => {
+    const fakeReq = {
+      originalUrl: '/firefox/whatever',
+      headers: {},
+    };
+    prefixMiddleWare(fakeReq, fakeRes, fakeNext, { _config: fakeConfig });
+    assert.equal(fakeRes.redirect.getCall(0).args[1], '/en-US/firefox/whatever');
+    assert.deepEqual(fakeRes.set.getCall(0).args[1], []);
```

Comment:
this would be a littler easier for me to read as `assert.deepEqual(fakeRes.set.getCall(0).args, ['vary', []])`